### ISSUE TITLE
[BUG] A bug related to ulysses_sequence_parallel_size and sglang sync rollout introduced from verion 0.4.1

### ISSUE THREAD

**Issue by twelveand0** [volcengine/verl#2639]
We have identified a bug related to the handling logic for `ulysses_sequence_parallel_size > 1` present from version *0.4.1* up to the latest version (commit id: 7fc3029a1ec407f6e56f1f1ff02a659071da3b1d). This bug causes a critical exception—specifically, a mismatch in the 1st dim size between `logits` and `labels` during the calculation of `logprobs_from_logits` —when using the configuration *fsdp + sglang + multi-turn + sync + sp > 1*.

Preliminary analysis suggests that this issue is related to the introduction of *multi_modal_inputs* (i.e., multi-modal input support) in version 0.4.1. In order to support *multi_modal_inputs*, the method `SGLangRollout::_req_level_generate_sequences` always returns a `non_tensor_batch` that includes the `multi_modal_inputs` key, without checking whether `data.return_multi_modal_inputs` is `True`. As a result, all requests are processed in the multi-modal branch when `logprobs_from_logits` is called, and the `input_ids_rmpad` are not properly split according to `ulysses_sequence_parallel_size (>1)`. Ultimately, this leads to a size mismatch in the first dimension between the `logits` and `labels` arguments passed to the `logprobs_from_logits` function, causing an exception.

## Faulty and Related Code

When upgrading from version 0.4.0 to 0.4.1, a new hyperparameter, **return_multi_modal_inputs**, was introduced to indicate whether there are multi-modal inputs  in the dataset or during the rollout process (the default value is `True`).

- ``return_multi_modal_inputs: False``: The dataset will not include the multi_modal_inputs field. This is recommended for multi-turn RL or tool-augmented rollouts, where the model may generate new multi-modal inputs dynamically during rollout, and you want to avoid conflicts or redundant data in the batch.

However, in the rollout code for processing each request in sglang’s synchronous mode (as shown below), whether or not to include the **multi_modal_inputs** non-tensor batch key in the return value is not determined based on whether `return_multi_modal_inputs` is set to `True`; instead, the **multi_modal_inputs** key is always included by default.

Related code until commit (id: 7fc3029a1ec407f6e56f1f1ff02a659071da3b1d): [https://github.com/volcengine/verl/blob/main/verl/workers/rollout/sglang_rollout/sglang_rollout.py#L1223](https://github.com/volcengine/verl/blob/main/verl/workers/rollout/sglang_rollout/sglang_rollout.py#L1223)

```python
    @GPUMemoryLogger(role="sglang rollout", logger=logger)
    @torch.no_grad()
    def _req_level_generate_sequences(self, prompts: DataProto, **kwargs) -> DataProto:
        """Generates multi-turn sequences for a batch of prompts.
        For multi-turn generation, each prompt is processed separately via
        `_req_level_generate_sequences` for better tool calling control.
        Note that in multi-turn generation, we repeat the prompts for rollout.n times in ray_trainer.
        Thus we do not need to repeat the prompts here and set the sampling parameter n to 1.
        """
        ......
        ......
        # free cache engine
        if self._engine is not None and self._tp_rank == 0:
            loop = asyncio.get_event_loop()
            loop.run_until_complete(self._engine.flush_cache())

        return DataProto(
            batch=batch,
            non_tensor_batch={
                "messages": np.array(messages),
                "reward_scores": np.array(reward_scores),
                "multi_modal_inputs": np.array(multi_modal_inputs, dtype=object),     // <====== HERE
            },
        )
```

This results in every sample being treated as having multi-modal inputs after the rollout stage—as long as the **multi_modal_input** key is present—even if `return_multi_modal_inputs=False` is set in the configuration. This behavior persists into the subsequent `compute_log_prob` stage. Relevant code includes [DataParallelPPOActor::compute_log_prob](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L291) and [DataParallelPPOActor::_forward_micro_batch](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L80).  The affected code is as follows:

[https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L138](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L138)

```python
                if self.use_ulysses_sp:
                    is_vlm_model = "multi_modal_inputs" in micro_batch.keys()              # <====== Here is affected code
                    if is_vlm_model:
                        # vlm model's inputs will be sliced after embedding
                        input_ids_rmpad, position_ids_rmpad, pad_size = ulysses_pad(
                            input_ids_rmpad,
                            position_ids_rmpad=position_ids_rmpad,
                            sp_size=self.ulysses_sequence_parallel_size,
                        )
                    else:
                        input_ids_rmpad, position_ids_rmpad, pad_size = ulysses_pad_and_slice_inputs(
                            input_ids_rmpad,
                            position_ids_rmpad=position_ids_rmpad,
                            sp_size=self.ulysses_sequence_parallel_size,
                        )
                    input_ids_rmpad_rolled, _, _ = ulysses_pad_and_slice_inputs(
                        input_ids_rmpad_rolled,
                        position_ids_rmpad=None,
                        sp_size=self.ulysses_sequence_parallel_size,
                    )

                input_ids_rmpad_rolled = input_ids_rmpad_rolled.squeeze(0)  # ((total_nnz / sp) + pad)

                # only pass input_ids and position_ids to enable flash_attn_varlen
                extra_args = {}
                if self.use_fused_kernels:
                    extra_args["temperature"] = temperature
                    extra_args["return_dict"] = True

                output = self.actor_module(
                    input_ids=input_ids_rmpad,
                    attention_mask=None,
                    position_ids=position_ids_rmpad,
                    **multi_modal_inputs,
                    use_cache=False,
                    **extra_args,
                )  # prevent model thinks we are generating

                if self.use_fused_kernels:
                    log_probs = output.log_probs.squeeze(0)  # (total_nnz,)
                    entropy_rmpad = output.entropy.squeeze(0)  # (total_nnz,)

                else:
                    logits_rmpad = output.logits.squeeze(0)  # (total_nnz, vocab_size)
                    logits_rmpad.div_(temperature)

                    # if use_sp: ((total_nnz / sp) + pad) ; if not use_sp: (batch, seqlen)
                    inplace_backward = True
                    if calculate_entropy:
                        inplace_backward = False
                    log_probs = logprobs_from_logits(
                        logits=logits_rmpad,
                        labels=input_ids_rmpad_rolled,
                        inplace_backward=inplace_backward,
                    )

```

When the hyperparameter `ulysses_sequence_parallel_size > 1` is set (i.e., sequence parallel is enabled), `self.use_ulysses_sp` in the above code will be `True`. As a result, the processing logic enters the **THEN** branch (see [dp_actor.py#L139](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L139)).

Since, as previously mentioned, every request is forced to include the **multi_modal_inputs** key after the rollout stage (in sglang + sync mode), the value of `is_vlm_model = "multi_modal_inputs" in micro_batch.keys()` will be `TRUE` at this point. As a result, execution will proceed to the *THEN* branch for `is_vlm_model`, which is specifically designed for handling multi-modal data. The key issue is that this branch calls the `ulysses_pad()` method, which does not slice `input_ids_rmpad` according to `sequence_parallel_size`; instead, the slicing is deferred to after the embedding stage. However, according to our configuration with `return_multi_modal_inputs=False`, both our input dataset and rollout process do not contain multi-modal inputs. In other words, we would expect the code here to go to the *ELSE* branch, where the `ulysses_pad_and_slice_inputs()` method is called to properly slice `input_ids_rmpad` according to `sequence_parallel_size`.

Additionally, regardless of whether `is_multi_model` is `TRUE` or not, the workflow will call the `ulysses_pad_and_slice_inputs()` method to slice `input_ids_rmpad_rolled` according to sequence_parallel_size. The sliced `input_ids_rmpad_rolled` will then be used as the `labels` argument in the subsequent [logprobs_from_logits](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L190) call.

In summary, at [dp_actor.py#L188](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L188), when calling the `logprobs_from_logits` function, the first argument `logits` and the second argument `labels` will differ in that one is sliced according to *sp size* while the other is not. As a result, when the execution enters `logprobs_from_logits >> logprobs_from_logits_flash_attn >> cross_entropy_loss`, it will trigger an assertion failure at `assert labels.shape == (nrows,)`.

### Root Cause Summary

Based on the above analysis, I believe the root cause is that in `SGLangRollout`, when the `_req_level_generate_sequences` method—which is responsible for processing each request—returns, it does not use the value of `return_multi_modal_inputs` to determine whether to include the `multi_modal_inputs` key in the return value.

## Trigger Conditions

**backend: FSDP**
**rollout_engine: sglang**
**rollout_mode: sync**
**ulysses_sequence_parallel_size: > 1**
**return_multi_modal_inputs**: False

```shell
python -m verl.trainer.main_ppo \
    --config-path="$CONFIG_PATH" \
    --config-name='gsm8k_multiturn_grpo' \
    algorithm.adv_estimator=grpo \
    data.train_batch_size=32 \
    data.max_prompt_length=163840 \
    data.max_response_length=32768 \
    data.filter_overlong_prompts=False \
    data.truncation='middle' \
    data.return_raw_chat=True \
    data.return_multi_modal_inputs=False \
    actor_rollout_ref.model.path=${BASE_MODEL} \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.optim.lr=2e-7 \
    actor_rollout_ref.actor.strategy=fsdp \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=24576 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.actor.fsdp_config.offload_policy=True \
    actor_rollout_ref.actor.fsdp_config.reshard_after_forward=True \
    actor_rollout_ref.actor.ulysses_sequence_parallel_size=8 \
    actor_rollout_ref.rollout.name=sglang \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.tensor_model_parallel_size=8 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.max_num_batched_tokens=196608 \
    actor_rollout_ref.rollout.max_model_len=196608 \
    actor_rollout_ref.rollout.mode=sync \
    actor_rollout_ref.rollout.n=4 \
    actor_rollout_ref.rollout.temperature=0.6 \
    actor_rollout_ref.rollout.top_k=20 \
    actor_rollout_ref.rollout.top_p=0.95 \
    actor_rollout_ref.ref.strategy=fsdp \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger='["console", "tensorboard"]' \
    trainer.project_name='gsm8k_async_rl' \
    trainer.experiment_name=${EXPERIMENT_NAME} \
    trainer.default_local_dir="${DEFAULT_LOCAL_DIR}/${EXPERIMENT_NAME}" \
    trainer.n_gpus_per_node=${n_gpus_per_node} \
    trainer.nnodes=${nnodes} \
    trainer.val_before_train=False \
    trainer.save_freq=1000 \
    trainer.test_freq=50 \
    data.train_files=train.parquet \
    data.val_files=test.parquet \
    actor_rollout_ref.rollout.multi_turn.tool_config_path="$PROJECT_DIR/examples/sglang_multiturn/config/tool_config/xxx.yaml" \
    actor_rollout_ref.rollout.multi_turn.max_user_turns=1 \
    actor_rollout_ref.rollout.multi_turn.max_assistant_turns=1 \
    actor_rollout_ref.rollout.multi_turn.max_parallel_calls=1 \
    actor_rollout_ref.rollout.multi_turn.max_tool_response_length=50000 \
    trainer.total_epochs=30 \
    2>&1 | tee "${EXPERIMENT_NAME}/run.log" $@
```

## Exception Stack

```python
output logprobs_from_logits_f lash_attn(logits,labels,inplace_backwardeinplace_backrard)
File "/workspace/verl-8.4.1/verl/utils/torch_functional.py", line 96, in logprobs_fron_logits_flash_attn
.
return self.logldecorated_function, wargs, *kwargs)
File "/workspace/vert-0.4.1/verl/utits/profiler/performance.py",Line 102.in log
File "/workspace/vert-0.4.1/verl/workers/actor/dp_actor-py",line 359,in compute_log.srob
File "/workspace/verl-0.4.1/verl/workers/actor/dp_actor.py",tine 119,in fonard_micro_batch
File "/workspace/verl-0.4.1/verl/utils/torch_functional.py", Line 86, in logprobs_from_logits
File "/opt/conda/Lib/python3.18/site-packages/flash_attn/eps/triton/cress_entrogy.py",lime 319,in cross_entrepy_less
File "/opt/conda/lib/python3.10/site-packages/flash_attn/ops/tritom/cross_entropy.py", line 170, in fonward
output = func(eargs, kwargs)
entrepy,log_probs self._fonward_micro_batch(
output cross_entropy_loss(legits,labels.inplace__)
File "/opt/conda/lib/python3.18/site-packages/torch/autograd/function.py", line 575, in apply
return CrossEntropyLoss.apply(
return super().apply(eargs, kwargs) # type: ignore[misc]
Set the environment variable HYORA_FULL_ERROR=1 for a complete stack trace.
log_probs = logprobs_from_logits(
assert labels,shape m (n_rows,)
AssertionError
main process return code:
```

## Patch Suggestion

When the `SGLangRollout::_req_level_generate_sequences` method returns the result, it should determine whether to include the `multi_modal_inputs` key in the return value based on the value of `return_multi_modal_inputs`.

**Comment by chenhaiq** [comment]
@nanjiangwill Can you have a look at this?

**Comment by nanjiangwill** [comment]
Thanks for the catch! i fixed it in #2667

**Comment by Kirrito-k423** [comment]
中文：感谢你提供详细的复现条件和根因分析。根据公开记录，这个问题已经由 #2667 修复并合并到 main：该 PR 明确标注 “Fix issue #2639”，合并提交为 f926dc90b000daab472f0fe13d9c8116211a2ea8。修复内容是让 SGLang sync rollout 先构造 non_tensor_batch，并且只在检测到多模态模型时加入 multi_modal_inputs，从而避免非多模态 / return_multi_modal_inputs=False 的请求被误判为 VLM 分支。基于这些证据，建议将此 issue 关闭为已修复。如果你在当前 main 或较新发布版本仍能复现，请提供新的 commit/version、完整配置和完整 traceback，维护者可以按回归问题继续跟进。

English: Thanks for the detailed reproduction conditions and root-cause analysis. Public records show that this was fixed and merged via #2667: the PR explicitly says “Fix issue #2639”, and the merge commit is f926dc90b000daab472f0fe13d9c8116211a2ea8. The fix builds non_tensor_batch first and only adds multi_modal_inputs when the rollout detects a multimodal model, which avoids routing non-multimodal / return_multi_modal_inputs=False requests into the VLM branch. Based on this evidence, this issue can be closed as fixed. If it still reproduces on current main or a newer release, please share the new commit/version, full config, and full traceback so maintainers can treat it as a regression.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
